File: preliminaryanalysis/Anish/Scripts/Spatial_distribution.py
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import geopandas as gpd
import contextily as ctx
from pathlib import Path


data = pd.read_csv(r"C:\ANISH-OLDLENOVO\Anish_School\CalCofi_capstone\capstone-calcofi-scripps\preliminaryanalysis\preliminarydata.csv")

meta_cols = {
    "", "unique.code", "S_C", "S_SC", "S_L", "S_S",
    "longitude", "latitude", "year", "season"
}

species_cols = [
    c for c in data.columns
    if c not in meta_cols
    and not c.startswith("Unnamed")
    and pd.api.types.is_numeric_dtype(data[c])
]


data["total_abundance"] = data[species_cols].fillna(0).sum(axis=1)

data["dominant_species"] = data[species_cols].idxmax(axis=1)
sizes = data["total_abundance"] / data["total_abundance"].max()
sizes = sizes * 200  # scale factor for visibility


gdf = gpd.GeoDataFrame(data, geometry = gpd.points_from_xy(data.longitude, data.latitude), 
                       crs="EPSG:4326")

gdf = gdf.to_crs(epsg =3857)


fig,ax = plt.subplots(figsize=(10,8))

scatter = ax.scatter(
    gdf.geometry.x,
    gdf.geometry.y,
    c=data["S_L"],
    s=sizes,
    cmap="viridis",
    alpha=0.7
)

ctx.add_basemap(ax, source=ctx.providers.CartoDB.Positron)

cbar = plt.colorbar(scatter, ax=ax)
cbar.set_label("CalCOFI Line (North to South)")

ax.set_title("Spatial Distribution of Standardized Fish Abundance")
ax.set_axis_off()

plt.tight_layout()
output_path = Path(__file__).parent / "spatial_distribution_abundance_python.png"
plt.savefig(output_path, dpi=300, bbox_inches="tight")

print(f"Figure saved to {output_path}")

plt.close()


#Below use same type of graph as above but now we identify top 10 dominant species on the map
species_totals = data[species_cols].sum()
top10_species = species_totals.sort_values(ascending=False).head(10).index

data["dominant_species"] = data[species_cols].idxmax(axis=1)

data["dominant_grouped"] = data["dominant_species"].where(
    data["dominant_species"].isin(top10_species),
    other="Other"
)

gdf = gpd.GeoDataFrame(data,geometry=gpd.points_from_xy(data.longitude, data.latitude), crs="EPSG:4326")

gdf = gdf.to_crs(epsg=3857)

sizes = gdf["total_abundance"] / gdf["total_abundance"].max()
sizes = sizes * 200

codes = gdf["dominant_grouped"].astype("category").cat.codes
categories = gdf["dominant_grouped"].astype("category").cat.categories

fig2, ax2 = plt.subplots(figsize=(10,8))

scatter2 = ax2.scatter(
    gdf.geometry.x,
    gdf.geometry.y,
    c=codes,
    s=sizes,
    cmap="tab20",
    alpha=0.8
)

ctx.add_basemap(ax2, source=ctx.providers.CartoDB.Positron)

handles = []
for i, species in enumerate(categories):
    handles.append(
        plt.Line2D([], [], marker='o', linestyle='',
                   color=plt.cm.tab20(i / len(categories)),
                   markersize=6,
                   label=species)
    )

ax2.legend(handles=handles,
           title="Dominant Species (Top 10)",
           bbox_to_anchor=(1.05, 1),
           loc="upper left",
           fontsize=8)

ax2.set_title("Spatial Distribution of Dominant Species")
ax2.set_axis_off()

plt.tight_layout()

output_path2 = Path(__file__).parent / "dominant_species_map_python.png"
plt.savefig(output_path2, dpi=300, bbox_inches="tight")
plt.close()

print(f"Dominant species map saved to {output_path2}")

File: preliminaryanalysis/Anish/Scripts/spatial_distribution.r
library(tidyverse)
library(sf)
library(ggspatial)
library(viridis)



data <- read_csv("preliminaryanalysis/preliminarydata.csv",
                 show_col_types = FALSE)



meta_cols <- c(
  "", "unique.code", "S_C", "S_SC", "S_L", "S_S",
  "longitude", "latitude", "year", "season"
)

species_cols <- setdiff(names(data), meta_cols)
species_cols <- species_cols[!grepl("^Unnamed", species_cols)]



data <- data %>%
  mutate(
    total_abundance = rowSums(select(., all_of(species_cols)), na.rm = TRUE)
  )

# Scale point sizes (ggplot uses small size units)
data <- data %>%
  mutate(size_scaled = (total_abundance / max(total_abundance)) * 3)

gdf <- st_as_sf(data,
                coords = c("longitude", "latitude"),
                crs = 4326)



map1 <- ggplot(gdf) +
  annotation_map_tile(type = "cartolight") +
  geom_sf(aes(size = size_scaled,
              color = as.numeric(S_L)),
          alpha = 0.7) +
  scale_color_viridis(option = "D") +
  coord_sf(crs = 3857) +
  labs(
    title = "Spatial Distribution of Standardized Fish Abundance",
    color = "CalCOFI Line (North → South)",
    size = "Relative Abundance"
  ) +
  theme_minimal()

ggsave(
  "preliminaryanalysis/Anish/visualizations/spatial_distribution_abundance.png",
  map1,
  width = 10,
  height = 8,
  dpi = 300
)



# Identify dominant species per station
data$dominant_species <- colnames(select(data, all_of(species_cols)))[
  max.col(select(data, all_of(species_cols)), ties.method = "first")
]

# Top 10 species overall
species_totals <- colSums(select(data, all_of(species_cols)), na.rm = TRUE)
top10_species <- names(sort(species_totals, decreasing = TRUE))[1:10]

data <- data %>%
  mutate(
    dominant_grouped = ifelse(dominant_species %in% top10_species,
                              dominant_species,
                              "Other")
  )

gdf2 <- st_as_sf(data,
                 coords = c("longitude", "latitude"),
                 crs = 4326)

map2 <- ggplot(gdf2) +
  annotation_map_tile(type = "cartolight") +
  geom_sf(aes(size = size_scaled,
              color = dominant_grouped),
          alpha = 0.8) +
  coord_sf(crs = 3857) +
  labs(
    title = "Spatial Distribution of Dominant Species",
    color = "Dominant Species (Top 10)",
    size = "Relative Abundance"
  ) +
  theme_minimal()

ggsave(
  "preliminaryanalysis/Anish/visualizations/dominant_species_map.png",
  map2,
  width = 10,
  height = 8,
  dpi = 300
)

cat("Both spatial maps saved in Anish folder.\n")